plugins/amqp_trace_plugin/include/eosio/amqp_trace_plugin/amqp_handler.hpp

```diff
@@ -0,0 +1,181 @@
+class amqp {
+public:
+   // called from amqp thread on errors
+   using on_error_t = std::function<void(const std::string& err)>;
+   // called from amqp thread on consume of message
+   // return true for ack, false for reject
+   using on_consume_t = std::function<bool(const char* buf, size_t s)>;
+
+   /// @param address AMQP address
+   /// @param name AMQP routing key
+   /// @param on_err callback for errors, called from amqp thread, can be nullptr
+   /// @param on_consume callback for consume on routing key name, called from amqp thread, null if no consume needed
+   amqp( const std::string& address, std::string name, on_error_t on_err, on_consume_t on_consume = nullptr )
+         : thread_pool_( "ampq", 1) // amqp is not thread safe, use only one thread
+         , name_( std::move( name ) )
+         , on_error_( std::move( on_err ) )
+         , on_consume_( std::move( on_consume ) )
+   {
+      AMQP::Address amqp_address( address );
+      ilog( "Connecting to AMQP address ${a} - Queue: ${q}...", ("a", std::string( amqp_address ))( "q", name_ ) );
+
+      handler_ = std::make_unique<amqp_handler>( *this, thread_pool_.get_executor() );
+      boost::asio::post( *handler_->amqp_strand(), [&]() {
+         connection_ = std::make_unique<AMQP::TcpConnection>( handler_.get(), amqp_address );
+      });
+      wait();
+   }
+
+   /// drain queue and stop thread
+   ~amqp() {
+      stop();
+   }
+
+   /// publish to AMQP address with routing key name
+   void publish( std::string exchange, std::string correlation_id, std::vector<char> buf ) {
+      boost::asio::post( *handler_->amqp_strand(),
+            [my=this, exchange=std::move(exchange), cid=std::move(correlation_id), buf=std::move(buf)]() {
+               AMQP::Envelope env( buf.data(), buf.size() );
+               env.setCorrelationID( cid );
+               my->channel_->publish( exchange, my->name_, env, 0 );
+      } );
+   }
+
+   /// publish to AMQP calling f() -> std::vector<char> on amqp thread
+   template<typename Func>
+   void publish( std::string exchange, std::string correlation_id, Func f ) {
+      boost::asio::post( *handler_->amqp_strand(),
+            [my=this, exchange=std::move(exchange), cid=std::move(correlation_id), f=std::move(f)]() {
+               std::vector<char> buf = f();
+               AMQP::Envelope env( buf.data(), buf.size() );
+               env.setCorrelationID( cid );
+               my->channel_->publish( exchange, my->name_, env, 0 );
+      } );
+   }
+
+   /// Explicitly stop thread pool execution
+   /// do not call from lambda's passed to publish or constructor e.g. on_error
+   void stop() {
+      if( handler_ ) {
+         // drain amqp queue
+         std::promise<void> stop_promise;
+         boost::asio::post( *handler_->amqp_strand(), [&]() {
+            stop_promise.set_value();
+         } );
+         stop_promise.get_future().wait();
+
+         thread_pool_.stop();
+         handler_.reset();
```

Comment:
This is scary, because it leaves the `TcpConnection` holding a dangling pointer, but it's safe (for now) because the first thing that `~TcpConnection` does is to clear its handler.